x/gamm/pool-models/balancer/pool.go

```diff
@@ -751,6 +764,49 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	return numShares, tokensJoined, nil
 }
 
+// CalcJoinPoolNoSwapShares calculates the number of shares created to execute an all-asset pool join with the provided amount of `tokensIn`.
+// The input tokens must contain exactly the same tokens as the pool contains
+//
+// Returns the number of shares created, the amount of coins actually joined into the pool, (in case of not being able to fully join),
+// and the remaining tokens in `tokensIn` after joining. If an all-asset join is not possible, returns an error.
+//
+// Since CalcJoinPoolNoSwapShares is non-mutative, the steps for updating pool shares / liquidity are
+// more complex / don't just alter the state.
+// We should simplify this logic further in the future using multi-join equations.
+func (p *Pool) CalcJoinPoolNoSwapShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, tokensJoined sdk.Coins, remainingTokens sdk.Coins, err error) {
+	// get all 'pool assets' (aka current pool liquidity + balancer weight)
+	poolAssetsByDenom, err := getPoolAssetsByDenom(p.GetAllPoolAssets())
+	if err != nil {
+		return sdk.ZeroInt(), sdk.NewCoins(), tokensIn, err
```

Comment:
I'm just going to point out which branches are not covered by tests.

@AlpinYukseloglu do you mind confirming please if it is possible to introduce test cases to cover these

This is the first branch

Reply:
Is it due to using a balancer pool constructor in `TestCalcJoinPoolNoSwapShares` where the constructor performs this validation?

What if we avoid it using it to trigger an invalid state and achieve full coverage?

We should be able to:
```
balancerPool := balancer.Pool{
  Address:            types.NewPoolAddress(defaultPoolId).String(),
  Id:                 defaultPoolId,
  PoolParams:         balancer.PoolParams{SwapFee: defaultSwapFee, ExitFee: defaultExitFee},
  PoolAssets:         test.poolAssets,
  FuturePoolGovernor: defaultFutureGovernor,
  TotalShares:        sdk.NewCoin(types.GetPoolShareDenom(defaultPoolId), types.InitPoolSharesSupply),
}
``` 

Now, we can force duplicate pool assets

Reply:
Please see the most recent case I added. It is still technically a trivial test case because duplicate pool assets are caught in multiple places throughout this function and ultimately end up in a division by zero even if you comment out all of them, but it should be covered now.